### ISSUE TITLE
never_loop does not understand break-with-label

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust-clippy#7397]
Lint name: never_loop

I tried this code:

```rust
            let response = 'outer: loop {
                for child in self.widgets.iter_mut() {
                    if id <= child.widget.id() {
                        let r = child.widget.send(mgr, id, child_event);
                        break 'outer (child.key.clone(), r);
                    }
                }
                debug_assert!(false, "SendEvent::send: bad WidgetId");
                return Response::Unhandled;
            };
```

(The code sample is [from here](https://github.com/kas-gui/kas/blob/e31b254df4e6663beef52c1a0d51ba386cf62be5/src/widget/view/list_view.rs#L552).)

I expected to see this happen: no report. I consider this a valid pattern as an (ugly) stand-in until (if ever) Rust gets for-break-with-default.

Instead, this happened:
```
error: this loop never actually loops
   --> src/widget/view/list_view.rs:552:28
```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b0 2021-06-17)
- `rustc -Vv`:
  ```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-unknown-linux-gnu
release: 1.53.0
LLVM version: 12.0.1
  ```

**Comment by Jarcho** [comment]
This particular case could be done with `find`:

```rust
let response = match self.widgets.iter_mut().find(|child| id <= child.widget.id()) {
    Some(child) => {
        let r = child.widget.send(mgr, id, child_event);
        (child.key.clone(), r)
    }
    None => {
        debug_assert!(false, "SendEvent::send: bad WidgetId");
        return Response::Unhandled;
    }
}
```

More generally something like this would work:

```rust
let mut iter = foo.iter();
let bar = loop {
    if let Some(bar) = iter.next() {
        if some_complex_thing() {
            break bar;
        }
    } else {
        // handle missing case here
    }
};
```

There's at least some debate as to which one is better in the general case, but if `find` works it should be used.

**Comment by dhardy** [comment]
> This particular case could be done with find:

I believe so. Semantically however `find` conveys the idea of *finding any matching candidate* whereas `for ...` conveys *do something in a certain order*. For this code that order is important, and while the code is equivalent, the semantics are still reason enough to use `find`.

Which is quite off topic here, the point merely being justification for using `for` over `find`.

The second example is a reasonable alternative, though may require explicit `drop(iter);` after the last line.

**Comment by camsteffen** [comment]
> I consider this a valid pattern as an (ugly) stand-in until (if ever) Rust gets for-break-with-default.

I don't think Clippy should make exceptions for such cases. You admitted it is ugly and Clippy did its job by agreeing with you. Whether the code should be changed or `#[allow]`ed is a matter of personal preference.

**Comment by HKalbasi** [comment]
This is a workaround for [RFC 2046](https://uta8a.github.io/rust-rfc-with-ja-dict/2046-label-break-value.html) and I'm not aware for a better and less ugly alternative. Maybe we can detect if tool chain is nightly, suggest block labels and if it is stable, don't emit warning (just for labeled loops with labeled breaks)?

**Comment by dhardy** [comment]
@HKalbasi code can't use nightly features if (as usual) stable compiler support is desired. And there are reasons beyond novel features to use a nightly compiler, e.g. [compile speed](https://bevyengine.org/learn/book/getting-started/setup/) and [proc-macro diagnostics](https://github.com/kas-gui/kas/#installation-and-dependencies).

**Comment by HKalbasi** [comment]
If it is not good we can disable this warning in nightly as well.

**Comment by oxalica** [comment]
I think we should allow label-break usages in replacement of `label_break_value`, since it's not yet stabilized. My use case is almost exact the same as the example of [RFC 2046 lable_break_value](https://uta8a.github.io/rust-rfc-with-ja-dict/2046-label-break-value.html). I think it's acceptable at least.

```rust
let some_value = 'got: loop {
    if let Some(...) = some_complex_condition() {
        break 'got some_value;
    }
    if some_other_condition() {
        return some_value; // Yes, I returns. So it's not easy to extract this loop into a function.
    }
    if let Ok(...) = some_extreme_condition() {
        break 'got some_value;
    }
    return otherwise;
};
```

**Comment by DesmondWillowbrook** [comment]
Also ran into this while looking at Rust Analyzer. Is there any progress on fixing this? I'd be happy to try my hand.

**Comment by xFrednet** [comment]
There isn't any progress AFAIK and nobody is assigned to the issue. So if you'd like to, you can claim it (`@rustbot claim`)

**Comment by DesmondWillowbrook** [comment]
The progress on https://github.com/rust-lang/rust/issues/48594#issuecomment-1201531228 seems to be nearly finished, so I don't think working on temporarily allowing a hacky workaround around a feature that seems it's going to be stabilized very soon seems like a good idea. I would be happy to revisit this in case the stabilization is delayed though.

**Comment by tommythorn** [comment]
I literally just hit this [1] and was appalled that Clippy complained *and* didn't offer a workaround.  The message "The loop is useless" is technically false as without the "loop" the code doesn't work.  I'm glad to see a better solution is forthcoming, but regardless the Clippy message is wrong and unhelpful.
[1] https://users.rust-lang.org/t/breaking-out-of-for/79893/5

**Comment by traviscross** [comment]
The problem here is that this is a *correctness* lint.  If we wanted to express that this pattern is simply *ugly*, then that would need to be a *style* lint (farcically, `clippy-hates-break-with-label`).  Even after `label-break-value` is stabilized in Rust 1.65, this correctness lint will remain *incorrect*.

People who write `#![allow(clippy::style)]` should be able to ignore Clippy's opinion about stylistic matters while benefiting from its other useful lints.

**Comment by DesmondWillowbrook** [comment]
Hmm, from what I understand it's labelled correctness as never looping in `for` loop (i.e. breaking out unconditionally) is far more likely to be an outright error (someone forgot to write the condition for breaking) instead of being a workaround for `label-break-value` not being stabilized yet,

Though now that it _is_ stabilized, it might be a good idea to add a properly worded suggestion for converting to a labelled block instead.

**Comment by DesmondWillowbrook** [comment]
This should now be fixed with #9837 and #9858.

**Comment by xFrednet** [comment]
Perfect. Thank you! Then I'll close this :)
